torch/csrc/distributed/autograd/context/context.cpp

```diff
@@ -108,15 +108,30 @@ void DistAutogradContext::addOutstandingRpc(
   outStandingRpcs_.push_back(futureMessage);
 }
 
-void DistAutogradContext::clearAndWaitForOutstandingRpcs() {
-  // Copy futures under lock, but wait for them outside the lock.
+std::shared_ptr<rpc::FutureMessage> DistAutogradContext::
+    clearAndWaitForOutstandingRpcsAsync() {
   std::unique_lock<std::mutex> lock(lock_);
   auto outStandingRpcs = std::move(outStandingRpcs_);
   lock.unlock();
 
-  for (const auto& outStandingRpc : outStandingRpcs) {
-    outStandingRpc->wait();
+  struct State {
+    explicit State(int32_t count)
+        : future(std::make_shared<rpc::FutureMessage>()), remaining(count) {}
+    std::shared_ptr<rpc::FutureMessage> future;
+    std::atomic<int32_t> remaining;
+  };
+  auto state = std::make_shared<State>(outStandingRpcs.size());
+  for (auto& rpc : outStandingRpcs) {
+    rpc->addCallback([state](const rpc::Message&) {
+      if (--state->remaining == 0) {
```

Comment:
callbacks can fire concurrently, do we need a lock here?

Reply:
no, this is an atomic, just need to decrement and process when one hits zero.